File: sorting/c-or-cpp/heap-sort.cpp
#include <iostream>
using namespace std;
void heapify(int a[], int n, int i);
void heapSort(int a[], int n);

// Driver code
int main()
{
    cout << "Enter the length of array" << endl;
    int n;
    cin >> n;
    int *a = new int(n);
    // Getting elements of array
    cout << "Enter the elements of array" << endl;
    for (int i = 0; i < n; i++)
        cin >> a[i];
    cout << "Original array:\n";
    for (int i = 0; i < n; i++)
        cout << a[i] << " ";
    heapSort(a, n);
    cout << "\nSorted array:\n";
    for (int i = 0; i < n; i++)
        cout << a[i] << " ";
    delete (a);
    return 0;
}

// To heapify a subtree rooted with node i which is an index in a[]
void heapify(int a[], int n, int i)
{
    int largest = i;   // Initialize largest as root
    int l = 2 * i + 1; 
    int r = 2 * i + 2; 

    // If left child is larger than root
    if (l < n && a[l] > a[largest])
        largest = l;

    // If right child is larger than largest so far
    if (r < n && a[r] > a[largest])
        largest = r;

    // If largest is not root
    if (largest != i)
    {
        swap(a[i], a[largest]);
        // Recursively heapify the affected sub-tree
        heapify(a, n, largest);
    }
}


void heapSort(int a[], int n)
{
    // Build heap (rearrange array)
    for (int i = n / 2 - 1; i >= 0; i--)
        heapify(a, n, i);
    // One by one extract an element from heap
    for (int i = n - 1; i > 0; i--)
    {
        // Move current root to end
        swap(a[0], a[i]);
        // call max heapify on the reduced heap
        heapify(a, i, 0);
    }
}

File: algorithms/JavaScript/src/sorting/heap-sort.js
function heapSort(arr){
    let arrayLength = arr.length;

    // make a max heap
    for (let index = Math.floor(arrayLength / 2)  - 1; index > -1; index--) {
        heapify(arr,arrayLength,index);
    }

    // put largest element begining of the array and extract it
    for (let index = arrayLength  - 1; index >= 0; index--) {
        let temp = arr[index];
        arr[index] = arr[0];
        arr[0] = temp;

        heapify(arr,index,0); // make max heap from rest of the array
    }

    return arr;
}

// making max heap function
function heapify(arr,arrayLength,index){
    let largest = index;
        left = 2 * index + 1;
        right = 2 * index + 2;
        needToSwap = false; 

    // if right is exist and it is bigger than root
    // make largest as right
    if(right < arrayLength && arr[largest] < arr[right]){
        largest = right;
        needToSwap = true;
    }

    // if left is exist and it is bigger than root
    // make largest as left
    if(left < arrayLength && arr[largest] < arr[left]){
        largest = left;
        needToSwap = true;
    }

    // swap
    if(needToSwap){
        let i = arr[index];
        arr[index] = arr[largest];
        arr[largest] = i;
        heapify(arr,arrayLength,largest);
    }
}

console.log(heapSort([100,99,999,1000,499,800,89,88,87,86,85,84,83,82,81,998,878,498,398,789,198,298,397,599,988]));